src/common/local_file_system.cpp

```diff
@@ -52,58 +52,48 @@ static void AssertValidFileFlags(uint8_t flags) {
 }
 
 #ifndef _WIN32
-bool LocalFileSystem::FileExists(const string &filename) {
-	if (!filename.empty()) {
-		if (access(filename.c_str(), 0) == 0) {
-			struct stat status;
-			stat(filename.c_str(), &status);
-			if (S_ISREG(status.st_mode)) {
-				return true;
-			}
-		}
-	}
-	// if any condition fails
-	return false;
-}
 
-bool LocalFileSystem::IsPipe(const string &filename) {
-	if (!filename.empty()) {
-		if (access(filename.c_str(), 0) == 0) {
-			struct stat status;
-			stat(filename.c_str(), &status);
-			if (S_ISFIFO(status.st_mode)) {
-				return true;
-			}
-		}
+static FileType GetFileTypeInternal(const struct stat &s) {
+	switch (s.st_mode & S_IFMT) {
+	case S_IFBLK:
+		return FileType::FILE_TYPE_BLOCKDEV;
+	case S_IFCHR:
+		return FileType::FILE_TYPE_CHARDEV;
+	case S_IFIFO:
+		return FileType::FILE_TYPE_FIFO;
+	case S_IFDIR:
+		return FileType::FILE_TYPE_DIR;
+	case S_IFLNK:
+		return FileType::FILE_TYPE_LINK;
+	case S_IFREG:
+		return FileType::FILE_TYPE_REGULAR;
+	case S_IFSOCK:
+		return FileType::FILE_TYPE_SOCKET;
+	default:
+		return FileType::FILE_TYPE_INVALID;
 	}
-	// if any condition fails
-	return false;
 }
 
-#else
-bool LocalFileSystem::FileExists(const string &filename) {
-	auto unicode_path = WindowsUtil::UTF8ToUnicode(filename.c_str());
-	const wchar_t *wpath = unicode_path.c_str();
-	if (_waccess(wpath, 0) == 0) {
-		struct _stati64 status;
-		_wstati64(wpath, &status);
-		if (status.st_mode & S_IFREG) {
-			return true;
+FileType LocalFileSystem::TryGetFileType(const std::string &path, optional_ptr<string> error) {
+	if (path.empty()) {
+		return FileType::FILE_TYPE_NOT_EXIST;
+	}
+	if (access(path.c_str(), 0) != 0) {
+		if (errno == ENOENT) {
+			return FileType::FILE_TYPE_NOT_EXIST;
 		}
+		error = make_uniq<string>(Exception::ConstructMessage("Cannot access path \"%s\": %s", path, strerror(errno)));
```

Comment:
For the reason stated earlier, this won't work
Should likely be `error = Exception::ConstructMessage(...);` once it's a `string &error`
At the call-site we just make a default initialized empty string and pass that in as a reference